sklearn/inspection/_permutation_importance.py

```diff
@@ -4,41 +4,37 @@
 from joblib import delayed
 
 from ..metrics import check_scoring
+from ..utils import Bunch
 from ..utils import check_random_state
 from ..utils import check_array
-from ..utils import Bunch
-
-
-def _safe_column_setting(X, col_idx, values):
-    """Set column on X using `col_idx`"""
-    if hasattr(X, "iloc"):
-        X.iloc[:, col_idx] = values
-    else:
-        X[:, col_idx] = values
-
-
-def _safe_column_indexing(X, col_idx):
-    """Return column from X using `col_idx`"""
-    if hasattr(X, "iloc"):
-        return X.iloc[:, col_idx].values
-    else:
-        return X[:, col_idx]
 
 
 def _calculate_permutation_scores(estimator, X, y, col_idx, random_state,
                                   n_repeats, scorer):
     """Calculate score when `col_idx` is permuted."""
-    original_feature = _safe_column_indexing(X, col_idx).copy()
-    temp = original_feature.copy()
+    random_state = check_random_state(random_state)
 
+    # Work on a copy of X to to ensure thread-safety in case of threading based
+    # parallelism. Furthermore, making a copy is also useful when the joblib
+    # backend is 'loky' (default) or the old 'multiprocessing': in those cases,
+    # if X is large it will be automatically be backed by a readonly memory map
+    # (memmap). X.copy() on the other hand is always guaranteed to return a
+    # writable data-structure whose columns can be shuffled inplace.
+    X_permuted = X.copy()
     scores = np.zeros(n_repeats)
+    shuffling_idx = np.arange(X.shape[0])
     for n_round in range(n_repeats):
-        random_state.shuffle(temp)
-        _safe_column_setting(X, col_idx, temp)
-        feature_score = scorer(estimator, X, y)
+        random_state.shuffle(shuffling_idx)
+        if hasattr(X_permuted, "iloc"):
+            # reset the index such that pandas reaffect by position instead of
+            # indices
+            X_permuted.iloc[:, col_idx] = X_permuted.iloc[
+                shuffling_idx, col_idx].reset_index(drop=True)
```

Comment:
What do you think of:

```py
if hasattr(X_permuted, "iloc"):
    X_index = X.index.copy()

for n_round in range(n_repeats):
    random_state.shuffle(shuffling_idx)
    if hasattr(X_permuted, "iloc"):
        col = X_permuted.iloc[shuffling_idx, col_idx]
        col.index = X_index
        X_permuted.iloc[:, col_idx] = col
```

Reply:
This version should handle case where `reset_index` does not match the original index. For example:

```py
X = pd.DataFrame({'a': [1, 2, 3], 'b': [3, 4, 5]}, index=['a', 'c', 'b'])
X_permuted = X.copy()

shuffling_idx = [2, 1, 0]
X_permuted.iloc[:, 0] = X_permuted.iloc[shuffling_idx, 0].reset_index(drop=True)

print(X_permuted)
    a  b
a NaN  3
c NaN  4
b NaN  5
```

With `col.index = `

```py
X = pd.DataFrame({'a': [1, 2, 3], 'b': [3, 4, 5]}, index=['a', 'c', 'b'])
X_index = X.index
X_permuted = X.copy()

col = X_permuted.iloc[shuffling_idx, 0]
col.index = X_index
X_permuted.iloc[:, 0] = col

print(X_permuted)
   a  b
a  3  3
c  2  4
b  1  5
```

Reply:
I updated the test to check for array equivalence on a DF with a str index and implemented your suggestion.